enhancements/ingress/set-delete-http-headers.md

```diff
@@ -0,0 +1,1031 @@
+
+Allowing the cluster-admin or route owner to override the `proxy` HTTP header isn't necessarily a security risk; it would only allow admins or route owners to inject configuration into their own applications, which they already control.  However, the `proxy` header is non-standard, and allowing it to be configured is potentially
+confusing.  For this reason, we do not allow setting the `proxy` header with this EP.
+
+#### Strict-Transport-Security
+Strict-transport-security is already handled separately, and we don't want to
+allow overrides of that configuration.
+
+If we allowed it, then route owners could impact each other's routes.
+
+It must be blocked so, we have put this
+header in the deny list.
+
+#### X-Forwarded-For
+`x-forwarded-for` and related headers were considered for the deny-list (as modifying these headers using this new API could affect configuration specified using
+another [option](https://docs.openshift.com/container-platform/4.12/networking/ingress-operator.html#nw-using-ingress-forwarded_configuring-ingress))
+but have actual use-cases (which aren't covered by the existing option). They will be allowed, with notes in the API and product documentation,
+and possibly a warning on admission.
+
+For example: If someone creates a Ingress Controller with the following:
+```sh
+apiVersion: operator.openshift.io/v1
+kind: IngressController
+metadata:
+  name: default
+  namespace: openshift-ingress-operator
+spec:
+  httpHeaders:
+    forwardedHeaderPolicy: Replace
+```
+Then the `set-header` stanzas will be added in `haproxy.config` using the [haproxy-config.template](https://github.com/openshift/router/blob/0cd9a58dff77127562f7851ccfa5cda79b8cef48/images/router/haproxy/conf/haproxy-config.template#L565-L580)
+However, if someone wants to set custom values for these then they need to use the API of this EP.
+
+If both are set then custom values will take over.  
+
+Please note that the setting of forwardedHeaderPolicy happens at the plain backend section of haproxy.config i.e. per route level and not at the front-end section
+of the haproxy.config level as per the above-mentioned link.
+
+Another use case is deleting the header which forwardedHeaderPolicy does not provide but this EP does.
+
+#### X-SSL
+Similarly to `x-forwarded-for`, there are legitimate use-cases for setting `x-ssl` and related headers, and allowing the cluster-admin or route owner to modify these headers presents no security risk.
+For example: HAProxy takes information from TLS handshake, tls session or the certificate that client provided. These headers are used for mtls and client certificates.
+
+These headers tell you whether the client provided a certificate, was the certificate valid, what was the content or cipher, distinguished name or expiration date
+on the certificate using converters with prefix like `ssl_` which can be used in the haproxy proxy config to take that information and put it http request headers which get sent
```

Comment:
I thought there were no other converters than the ones listed.  What converters are you talking about here?
```suggestion
These headers tell you whether the client provided a valid certificate, and certificate values such as content,  cipher, distinguished name, or expiration date on the certificate.  

```

Reply:
The original text looks good to me. or is used correctly from my understanding.  It is just to explain why ssl related fetchers can't be an issue.

Reply:
i have made some changes.

Reply:
> I thought there were no other converters than the ones listed. What converters are you talking about here?

HAProxy has scores of converters, documented here: <http://cbonte.github.io/haproxy-dconv/2.6/configuration.html#7.3.1>.  The reason why this enhancement doesn't allow more converters is explained in "Should this enhancement allow setting dynamic header values?" under "Open Questions".